### PR TITLE
Hexbins

### PR DESCRIPTION
**Pull request by royrutto**
*TODO*
- [x] Bin highlighted points on view-by
- [x] Change Bin size on zoom

### CODE REVIEW

File: src/hatti/map/utils.cljs [hunk 1]
```diff
@@ -506,6 +506,87 @@
                                   :line-cap "round"}}
     :else {:layer-type "circle" :style :point}))
 
+(defn to-fixed
+  [n number-of-decimal-places]
+  (when n
```
Comment (pld) [inline src/hatti/map/utils.cljs]: can you use `some->` here to combine the `when` and `->`

File: src/hatti/map/utils.cljs [hunk 2]
```diff
@@ -506,6 +506,87 @@
...
+  [map layer-id & [geojson]]
+  (let [get-rendered-features #(.queryRenderedFeatures
+                                map (clj->js {:layers [layer-id]}))
+        rendered-features (or geojson
+                              (clj->js
+                               {:type "FeatureCollection"
+                                :features (get-rendered-features)}))
+        ;; Get bounding box for rendered features
+        bbox (.bbox js/turf (clj->js rendered-features))
+        cellWidth 60
```
Comment (pld) [inline src/hatti/map/utils.cljs]: where does this number come from? should either be an arg or in a descriptive def

File: src/hatti/map/utils.cljs [hunk 3]
```diff
@@ -506,6 +506,87 @@
...
+    (assoc hexbins :features features-w-count
+           :properties {:total
+                        total})))
+
+(defn show-hexbins
+  "Renders hexbin layer on map."
+  [map id_string geojson]
+  (let [id "hexgrid"
+        hexgrid (generate-hexgrid map id_string geojson)
+        _ (.log js/console (clj->js hexgrid))
```
Comment (pld) [inline src/hatti/map/utils.cljs]: remove

File: src/hatti/map/utils.cljs [hunk 4]
```diff
@@ -506,6 +506,87 @@
...
+        total (-> hexgrid :properties :total)]
+    (add-mapboxgl-source map id {:geojson hexgrid})
+    (add-mapboxgl-layer map id
+                        "fill"
+                        :paint {:fill-outline-color
+                                {:property "point_count"
+                                 :stops [[0 "transparent"]
+                                         [100 "#ccc"]]}
```
Comment (pld) [inline src/hatti/map/utils.cljs]: can this, and colors on line 579 and 580 not be controlled by stylesheets? If not these colors should also go to args or descriptive defs

File: src/hatti/map/utils.cljs [hunk 5]
```diff
@@ -519,11 +600,14 @@
...
       (set-mapboxgl-paint-property
        map id_string (get-style-properties style :normal :stops stops))
+      ;; add layer that acts as border for point type layers,
+      ;; otherwise remove the layer if it exists
       (if (= :point style)
         (add-mapboxgl-layer map id_string layer-type
                             :layer-id circle-border
                             :paint {:circle-color "#fff" :circle-radius 6})
         (when (.getLayer map circle-border) (.removeLayer map circle-border)))
+      ;;(show-hexbins map id_string geojson)
```
Comment (pld) [inline src/hatti/map/utils.cljs]: is a comment? add space after ;;, remove parens. if it's commented code think we can remove

File: src/hatti/views/map.cljs [hunk 6]
```diff
@@ -476,7 +429,7 @@
         (let [with-suffix #(if-not (om/get-state owner :expanded) %
                                    (str % " leaflet-control-layers-expanded"))]
           (html
-           [:div.leaflet-left.leaflet-bottom {:style {:margin-bottom "148px"}}
+           [:div.leaflet-left.leaflet-bottom {:style {:margin-bottom "190px"}}
```
Comment (pld) [inline src/hatti/views/map.cljs]: this is old code, but this definitely should be moved to stylesheets and removed from here

File: src/hatti/views/map.cljs [hunk 7]
```diff
@@ -489,6 +442,27 @@
...
+  (reify
+    om/IInitState
+    (init-state [_]
+      {:active? false})
+    om/IRenderState
+    (render-state [_ {:keys [active?]}]
+      "Render layer selector component w/css + expansion technique from
+      leaflet layer control."
+      (html
+       [:div.leaflet-left.leaflet-bottom {:style {:margin-bottom "145px"}}
```
Comment (pld) [inline src/hatti/views/map.cljs]: move to stylesheets

File: src/hatti/map/utils.cljs [hunk 8]
```diff
@@ -506,6 +508,84 @@
...
+           :properties {:min-count (apply min point-counts)
+                        :max-count (apply max point-counts)})))
+
+(defn show-hexbins
+  "Renders hexbin layer on map."
+  [map id_string geojson opts]
+  (let [id "hexgrid"
+        hexgrid (generate-hexgrid map id_string geojson opts)
+        {:keys [min-count max-count]} (:properties hexgrid)
+        max-color (or (:cell-color opts) "#08519c")]
```
Comment (pld) [inline src/hatti/map/utils.cljs]: move hex colors to a def

File: src/hatti/map/utils.cljs [hunk 9]
```diff
@@ -506,6 +508,84 @@
...
+      (add-mapboxgl-source map id {:geojson hexgrid})
+      (add-mapboxgl-layer map id
+                          "fill"
+                          :paint {:fill-outline-color
+                                  {:property "point-count"
+                                   :stops [[0 "transparent"]
+                                           [max-count "#ccc"]]}
+                                  :fill-color {:property "point-count"
+                                               :stops [[0 "transparent"]
+                                                       [min-count "#eff3ff"]
```
Comment (pld) [inline src/hatti/map/utils.cljs]: move hex colors to a def

File: src/hatti/map/utils.cljs [hunk 10]
```diff
@@ -506,6 +508,84 @@
...
+        {:keys [min-count max-count]} (:properties hexgrid)
+        max-color (or (:cell-color opts) "#08519c")]
+    (when (and min-count max-count)
+      (add-mapboxgl-source map id {:geojson hexgrid})
+      (add-mapboxgl-layer map id
+                          "fill"
+                          :paint {:fill-outline-color
+                                  {:property "point-count"
+                                   :stops [[0 "transparent"]
+                                           [max-count "#ccc"]]}
```
Comment (pld) [inline src/hatti/map/utils.cljs]: move hex colors to a def

File: src/hatti/views/map.cljs [hunk 11]
```diff
@@ -439,25 +387,47 @@
...
+                          (mu/as-geojson new-map-data flat-form new-field)
+                          geojson)
+            view-by-changed? (not= old-viewby new-viewby)
+            cell-width-changed (not= old-cell-width new-cell-width)
+            opts (when (and show-hexbins? view-by-changed?)
+                   (vb/get-selected-ids new-viewby))
+            hexbin-opts (assoc opts :cell-width new-cell-width)]
+        ;; Update layers if data changes
+        (when data-changed?
+          (when (and (not-empty new-field) (not= geojson new-geojson))
```
Comment (pld) [inline src/hatti/views/map.cljs]: can this be combined w/the outer `when`?

File: src/hatti/views/map.cljs [hunk 12]
```diff
@@ -489,6 +475,85 @@
...
+    (init-state [_]
+      {:slider-value 0
+       :zoom->bin? true
+       :value->size {0 500 1 250 2 100 3 50 4 25 5 10 6 5 7 1 8 0.5}})
+    om/IWillReceiveProps
+    (will-receive-props [_ next-props]
+      (when (om/get-state owner :zoom->bin?)
+        (when-let [map (-> next-props :map-page :mapboxgl-map)]
+          (let [zoom (.getZoom map)
+                value (cond
```
Comment (pld) [inline src/hatti/views/map.cljs]: can be written as

```clojure
(condp #(<= %2 %1) zoom
  2 0
  5 1
  8 2
  11 3
  14 4
  17 5
  20 6)
```

but that will raise an exception for `> 20` and return 0 for `< 0`.

```clojure
(condp #(<= %2 %1) zoom
  -1 nil
  2  0
  5  1
  8  2
  11 3
  14 4
  17 5
  20 6
  nil)
```

will exactly duplicate the existing functionality

File: src/hatti/views/map.cljs [hunk 13]
```diff
@@ -489,6 +475,85 @@
...
+                                   (let
+                                    [v (js/parseInt (.. e -target -value))]
+                                     (om/update-state!
+                                      owner #(assoc % :zoom->bin? false
+                                                    :slider-value v))))
+               below-1km? (> 1 cell-size)]
+           [:div.map-overlay.mapboxgl-ctrl-bottom-left
+            [:div.map-overlay-inner
+             [:label.slider "Cell Width: "
+              [:span#slider-value (str (if below-1km?
```
Comment (pld) [inline src/hatti/views/map.cljs]: why not always convert to meters?

Comment (pld) [inline src/hatti/views/map.cljs]: oh, this gets displayed to users? that makes sense then

Comment (pld) [inline src/hatti/views/map.cljs]: you can write that `if` more concisely as:

```clojure
(cond-> cell-size below-1km? (* 1000))
```

File: src/hatti/map/utils.cljs [hunk 14]
```diff
@@ -506,6 +508,88 @@
...
+                                                       [min-count min-color]
+                                                       [max-count max-color]]}
+                                  :fill-opacity 0.5}))))
+
+(defn remove-hexbins
+  "Remove hexbins layer from map"
+  [map]
+  (let [id "hexgrid"]
+    (when (.getLayer map id) (.removeSource map id))
+    (when (.getLayer map id) (.removeLayer map id))))
```
Comment (pld) [inline src/hatti/map/utils.cljs]: can't this and the above can be combined as

```clojure
(when (.getLayer map id)
  (.removeSource map id)
  (.removeLayer map id))

File: src/hatti/views/map.cljs [hunk 15]
```diff
@@ -439,25 +392,58 @@
...
+            view-by-changed? (and (not= old-viewby new-viewby)
+                                  (not (nil? old-viewby)))
+            cell-width-changed? (not= old-cell-width new-cell-width)
+            opts (when (and show-hexbins? view-by-changed?)
+                   (vb/get-selected-ids new-viewby))
+            hexbin-opts (assoc opts :cell-width new-cell-width)]
+        ;; Update layers if data changes
+        (when (and data-changed? (not-empty new-field)
+                   (not= geojson new-geojson))
+          (when (.getLayer mapboxgl-map layer-id)
```
Comment (pld) [inline src/hatti/views/map.cljs]: Isn't this the same as https://github.com/onaio/hatti/pull/174/files#diff-48c0c62913167da8085340a83a0734d6R591, helper function?

File: src/hatti/views/map.cljs (hunk 13 above)
Comment (royrutto) [inline src/hatti/views/map.cljs]: yes it get's displayed to users.

**Comment by royrutto** [comment]
@pld this can be reviewed again. 

The tests require more work because the JS dependencies required to run tests on the map view are in zebra (MapboxGL and turf-js) and we need to set up a reader for the fixtures in CLJS.

**Pull request merged**